### PR TITLE
Optimise GetLog API query in v1alpha3 Logs API

### PR DESCRIPTION
**Pull request by khrm**
In logql queries, optimised the following:

* Duration where we search should be small.
* Searching for uid(TaskRun/PipelineRunUID) after parsing data.

Also, we are caching Token.

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you review them:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] [Tested your changes locally](https://github.com/tektoncd/results/blob/main/docs/DEVELOPMENT.md) (if this is a code change)
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user-facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contain the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

### CODE REVIEW

**Review by khrm — COMMENTED** [review]
/kind cleanup

**Comment by khrm** [comment]
I have added the documentation for new parameter.

File: pkg/api/server/v1alpha2/plugin_logs.go [hunk 1]
```diff
@@ -105,25 +163,41 @@ func (s *LogPluginServer) getLokiLogs(writer *logs.BufferedLog, parent, id strin
 		return status.Error(codes.Internal, "Error streaming log")
 	}
 
+	if resp.StatusCode != http.StatusOK {
+		s.logger.Errorf("Loki API request failed with status code: %d", resp.StatusCode)
```
Comment (enarha) [inline pkg/api/server/v1alpha2/plugin_logs.go:167]:
```suggestion
		s.logger.Errorf("Loki API request failed with HTTP status code: %d", resp.StatusCode)
```

Comment (enarha) [reply pkg/api/server/v1alpha2/plugin_logs.go:167]: This is the differentiate between the HTTP request error and the Loki response error below.

File: pkg/api/server/v1alpha2/server.go [hunk 2]
```diff
@@ -79,13 +74,17 @@ type LogPluginServer struct {
...
 	staticLabels       string
-	MaxRetention       time.Duration
 
 	config *config.Config
 	logger *zap.SugaredLogger
 	auth   auth.Checker
 	db     *gorm.DB
 	client *http.Client
+
+	forwarderBufferDuration time.Duration
```
Comment (enarha) [inline pkg/api/server/v1alpha2/server.go:84]: If you decide to update the configuration option name as I suggested above, you can change the var name to match (e.g forwarderDelayDuration).

Comment (khrm) [reply pkg/api/server/v1alpha2/server.go]: I have changed it to forwarderDelayDuration.

File: pkg/api/server/v1alpha2/plugin_logs.go [hunk 3]
```diff
@@ -60,41 +68,91 @@ func (s *LogPluginServer) GetLog(req *pb3.GetLogRequest, srv pb3.Logs_GetLogServ
...
+			return err
+		}
+		endTime = strconv.FormatInt(data.Status.CompletionTime.Add(s.forwarderBufferDuration).UTC().Unix(), 10)
+
+	case typeTaskRun:
+		uidKey = taskRunUIDKey
+		data := &pipelinev1.TaskRun{}
+		err := json.Unmarshal(rec.Data, data)
+		if err != nil {
+			s.logger.Errorf("failed to marshal pipelinerun data for fetching log, err: %s:", err.Error())
```
[the remark this answers is not in the document]
Comment (khrm) [reply pkg/api/server/v1alpha2/plugin_logs.go]: Yes, we should.

[the remark this answers is not in the document]
Comment (khrm) [reply pkg/api/server/v1alpha2/plugin_logs.go]: Changed.

File: pkg/api/server/v1alpha2/plugin_logs.go [hunk 4]
```diff
@@ -60,41 +68,91 @@ func (s *LogPluginServer) GetLog(req *pb3.GetLogRequest, srv pb3.Logs_GetLogServ
...
 
-	now := time.Now()
+	var startTime, endTime, uidKey string
+	switch rec.Type {
+	case typePipelineRun:
+		uidKey = pipelineRunUIDKey
+		data := &pipelinev1.PipelineRun{}
+		err := json.Unmarshal(rec.Data, data)
+		if err != nil {
+			s.logger.Errorf("failed to marshal pipelinerun data for fetching log, err: %s:", err.Error())
```
[the remark this answers is not in the document]
Comment (khrm) [reply pkg/api/server/v1alpha2/plugin_logs.go]: Yes, changed.

File: docs/logging-support.md [hunk 5]
```diff
@@ -25,5 +25,4 @@ These are the common configuration options for all third party logging APIs.
 - `LOGGING_PLUGIN_PROXY_PATH`: The path to the proxy to use for the third party logging API. (optional)
 - `LOGGING_PLUGIN_CA_CERT`: The CA certificate to use for the third party logging API. This should ideally be passed as environment variable in the deployment spec of the results-api pod. (optional)
 - `LOGGING_PLUGIN_TLS_VERIFICATION_DISABLE`: Set to `true` to disable TLS verification for the third party logging API. (optional)
-
-Also, `MAX_RETENTION` is passed to the results API from the Retention Policy ConfigMap.
+- `LOGGING_PLUGIN_FORWARDER_BUFFER_DURATION`: Buffer Duration is the max duration taken by third party logging system to forward and store the logs after completion of taskrun and pipelinerun. This is used to search between start time of runs and completion plus buffer duration.
```
[the remark this answers is not in the document]
Comment (khrm) [reply docs/logging-support.md]: Changed to `LOGGING_PLUGIN_FORWARDER_DELAY_DURATION`.

**Comment by khrm** [comment]
Testing these changes: https://github.com/tektoncd/results/compare/79c2888413f7b3866b91e157872420f81e915881..33890fd2ece992e4a8276fe3581344f396b1459e

**Review by enarha — COMMENTED** [review]
/lgtm

**Comment by khrm** [comment]
Done with testing: 
CLI
image
GUI
image

**Pull request merged**